compiler/rustc_lint/src/types.rs

```diff
@@ -1084,7 +1210,67 @@ impl<'a, 'tcx> ImproperCTypesVisitor<'a, 'tcx> {
                 FfiSafe
             }
 
-            ty::RawPtr(ty, _) | ty::Ref(_, ty, _) => self.check_type_for_ffi(acc, ty),
+            ty::RawPtr(inner_ty, _) | ty::Ref(_, inner_ty, _) => {
+                if let TypeSizedness::UnsizedBecauseForeign | TypeSizedness::Sized =
+                    get_type_sizedness(self.cx, inner_ty)
+                {
+                    // there's a nuance on what this lint should do for function definitions
+                    // (`extern "C" fn fn_name(...) {...}`) versus declarations (`extern "C" {fn fn_name(...);}`).
+                    // (this is touched upon in https://github.com/rust-lang/rust/issues/66220
+                    // and https://github.com/rust-lang/rust/pull/72700)
+                    //
+                    // The big question is: what does "ABI safety" mean? if you have something translated to a C pointer
+                    // (which has a stable layout) but points to FFI-unsafe type, is it safe?
+                    // on one hand, the function's ABI will match that of a similar C-declared function API,
+                    // on the other, dereferencing the pointer in not-rust will be painful.
+                    // In this code, the opinion is split between function declarations and function definitions.
+                    // For declarations, we see this as unsafe, but for definitions, we see this as safe.
+                    // This is mostly because, for extern function declarations, the actual definition of the function is written somewhere else,
+                    // so the fact that a pointer's pointee should be treated as opaque to one side or the other can be explicitely written out.
+                    // For extern function definitions, however, both callee and some callers can be written in rust,
+                    // so developers need to keep as much typing information as possible.
```

Comment:
I'll reply here once rather than once per comment on that commented explanation.

Yeah, the explanation I used here was very wishy-washy. Thankfully I have a better one now.
Namely, the compiler (very likely?) needs you to have non-opaque types in the function header if you are going to use them in a non-opaque way in the function body)